docs/design/persistent-volume-provisioning.md

```diff
@@ -0,0 +1,159 @@
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Persistent Volume Provisioning
+
+This document proposes a model for dynamically provisioning
+
+### tl;dr
+
+One new API kind:
+
+A `PersistentVolumeController` (PVC) is a storage resource provisioned by an administrator.  PVCs do not have a namespace.  Just as a `ReplicationController` maintains a number of replicas of a pod, a `PersistentVolumeController` maintains a minimum number of replicas of a `PersistentVolume`.  A PVC creates new volumes from a template up to a maximum replica count in increments of the minimum replica count.  A well-provisioned cluster will have many PVCs.
+
+One new system component:
+
+`PersistentVolumeControllerManager` is a singleton control loop running in master that manages all PVControllers in the system.  The PVCM reconciles the current supply of available PersistentVolumes in the system with the desired levels according to the PVControllers.   This process is similar to the ```ReplicationManager``` that manages ReplicationControllers.
+
+Three new volume plugin interfaces:
+
+* Recycler -- knows how to scrub a volume clean so it can become available again as a resource
+* Creator -- create new instances of a PV from a template.
+* Deleter -- deletes instances of a PV and allows the plugin to determine how to remove it from the underlying infrastructure
+
+Volume plugins can implement any applicable interfaces.  Each plugin will document its own support for dynamic provisioning.
+
+
+### Goals
+
+* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
+* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
+
+
+#### API types
+
+```
+// PersistentVolumeController represents the configuration of a persistent volume.
+type PersistentVolumeController struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired specification of this PV controller.
+	Spec PersistentVolumeControllerSpec `json:"spec,omitempty"`
+
+	// Status is the current status of this PV controller.
+	Status PersistentVolumeControllerStatus `json:"status,omitempty"`
+}
+
+// PersistentVolumeControllerSpec is the specification of a PV controller.
+// The controller creates volumes exactly like the Template until the unbound count equals the replica count
+type PersistentVolumeControllerSpec struct {
+	// MinimumReplicas is the minimum number of unbound persistent volumes of this type desired in the system maintain
+	MinimumReplicas int `json:"minimumReplicas"`
+
+	// MaximumReplicas is the maximum total number of persistent volumes desired in the system
+	MaximumReplicas int `json:"maximumReplicas"`
+	
```

Comment:
Hmm for what it is worth I found the above terms confusing at first pass. How about s/minimum/minimumUnbound ?
or just "replicas" to use the same terminology as replication controllers. maximumReplicas refers to the total of bound and unbound replicas. Correct ?